Sprint 2

sgp900 opened the pull request.

sgp900 committed:
sprint-2

sgp900 committed:
remove parasite vscode folder

podlovchenko commented on package.json:
```diff
@@ -3,19 +3,31 @@
     "@anikin/parcel-transformer-handlebars-precompile": "1.0.1",
     "@parcel/transformer-sass": "2.7.0",
-    "parcel": "2.7.0"
+    "@typescript-eslint/eslint-plugin": "^5.44.0",
```
Лучше задавать фиксированные версии зависимостям. Если однажды решишь, удалить package-lock.json файл и попытаться повторно установить зависимости с помощью команды npm install, npm установит последние версии зависимостей, что может привести к ошибкам

podlovchenko commented on package.json:
```diff
@@ -3,19 +3,31 @@
+    "jslint": "eslint ./src/**/*.ts",
+    "dev": "parcel ./src/index.html --port 3000",
+    "build": "parcel build ./src/index.html",
+    "start": "parcel build ./src/index.html && node ./server.js"
```
Можно переиспользовать 
```suggestion
    "start": "npm run build && node ./server.js"
```

podlovchenko commented on src/components/Button/button.ts:
```diff
@@ -0,0 +1,25 @@
+
+export default class Button extends Block<ButtonProps> {
+  constructor(props: ButtonProps) {
+    if (props.type === undefined) {
+      // eslint-disable-next-line no-param-reassign
+      props.type = "submit";
+    }
+    super(props);
```
Можно
```suggestion
    super({
    type: 'submit',
    ...props
    });
```

podlovchenko commented on src/components/ChatList/chatList.ts:
```diff
@@ -0,0 +1,40 @@
+  chats: ChatProps[];
+}
+
+export default class ChatList extends Block<ChatsListProps> {
```
Использование дефолтных экспортов, на мой взгляд, это не очень хорошая практика.

Они позволяют задавать разные имена импортируемым объектам, из-за этого страдает читабельность, экземпляры с разными именами очень трудно потом искать по проекту. Так же затрудняется рефакторинг таких импортов.

Если задавать обычный экспорт, мы можем быть почти везде точно уверены, что имена используются везде одинаковые и IDE в таком случае спокойно их сможет поменять в случае рефакторинга.

Хорошая статья на эту тему
https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/

podlovchenko commented on src/components/CurrentChat/currentChat.ts:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  private static createMessages(props: MessageProps[]) {
+    return props.map((data) => new Message({ ...data }));
```
```suggestion
    return props.map((data) => new Message(data));
```

podlovchenko commented on src/components/CurrentChat/template.hbs:
```diff
@@ -0,0 +1,96 @@
+              <button
+                onclick="let overlay = document.querySelector('.overlay.add-user');
+                 overlay.classList.toggle('hide');"
+              ><div
+                  class="btn-drop__drop-icon currentchat-action__icon currentchat-action__icon_add"
+                ></div><p>Добавить пользователя</p></button>
```
Что-то поехало форматирование, такое сложно читать и поддерживать, может помочь prettier https://prettier.io/docs/en/install.html

podlovchenko commented on src/components/Input/input.ts:
```diff
@@ -0,0 +1,23 @@
+  placeholder?: string;
+  class: string;
+  data?: string;
+  // eslint-disable-next-line no-unused-vars
```
Это выключение проверки здесь излишне

podlovchenko commented on src/components/Message/style.scss:
```diff
@@ -0,0 +1,17 @@
+    align-self: flex-end;
+  }
+
+  a {
```
Лучше использовать классы, а не теги
Статья на эту тему https://frontstuff.io/you-need-to-stop-targeting-tags-in-css

podlovchenko commented on src/utils/HTTPTransport.ts:
```diff
@@ -0,0 +1,98 @@
+}
+
+type Options = {
+  method?: string;
```
Можно определить метод через enum и типизировать здесь

podlovchenko commented on src/utils/HTTPTransport.ts:
```diff
@@ -0,0 +1,98 @@
+
+type Options = {
+  method?: string;
+  headers?: Object;
```
```suggestion
  headers?: Record<string, string>
```

podlovchenko commented on src/utils/block.ts:
```diff
@@ -0,0 +1,268 @@
+    });
+  }
+
+  _componentDidUpdate(oldProps: P, newProps: P) {
```
Для приватных методов не хватает модификатора private

podlovchenko commented on tsconfig.json:
```diff
@@ -0,0 +1,108 @@
+{
+  "compilerOptions": {
+    /* Visit https://aka.ms/tsconfig to read more about this file */
+
+    /* Projects */
+    // "incremental": true,                              /* Save .tsbuildinfo files to allow for incremental compilation of projects. */
+    // "composite": true,                                /* Enable constraints that allow a TypeScript project to be used with project references. */
+    // "tsBuildInfoFile": "./.tsbuildinfo",              /* Specify the path to .tsbuildinfo incremental compilation file. */
+    // "disableSourceOfProjectReferenceRedirect": true,  /* Disable preferring source files instead of declaration files when referencing composite projects. */
+    // "disableSolutionSearching": true,                 /* Opt a project out of multi-project reference checking when editing. */
+    // "disableReferencedProjectLoad": true,             /* Reduce the number of projects loaded automatically by TypeScript. */
+
+    /* Language and Environment */
+    "target": "es2016" /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */,
+    "lib": [
+      "ESNext",
+      "DOM"
+    ] /* Specify a set of bundled library declaration files that describe the target runtime environment. */,
```
По заданию должны быть следующие настройки
```
"noImplicitAny": true,

"noUnusedLocals": true,

"noUnusedParameters": true,

"sourceMap": true,

"strictNullChecks": true
```

podlovchenko reviewed (changes requested):
Добрый вечер, есть несколько замечаний
1. Нет настроек в tsconfig по заданию
2. Много где выключаются проверки eslint, стоит настроить eslint так, чтобы не приходилось это
3. Было бы здорово почистить проект от ненужного закомментированного кода

sgp900 committed:
1. В конфигурационном файле TypeScript явно включены следующие настройки:
 - "noImplicitAny": true,
 - "noUnusedLocals": true,
 - "noUnusedParameters": true,
 - "sourceMap": true,
 - "strictNullChecks": true

 2. Приватным методам добавлен модификатор private;

 3. В классе HTTPTransport:
  - константа METHODS переопределена через enum;
  - измен тип Options;

 4. Убрана стилизация по тегам.

 5. Перенастроен ESLint, чтобы избежать засорение кода исключениями

 6. Дефолтный экспорт классов заменён на явный

 7. В классе Button дефолтные значения параметров "type" и "class" задаются с помощью деструктуризации объекта props.

 8. В файле package.json
  - пере использован скрипт build в start;
  - использованы фиксированные версии пакетов;

 9. Код почищен от закомментированного кода, убрана лишняя деструктуризация объектов.

 10. Добавлен prettier 2.4.1.

podlovchenko commented:
Спасибо за оперативные правки, работу принимаю)

sgp900 committed:
1. В компоненты ProfileField и PlaceHolderInput добавлен children - ErrorBox.
Чтобы избежать проблем с потерей фокуса при перерисовке компонента.

2. В классе Block добавлен метод _removeEvents. Чтобы избежать утечки памяти, при перерисовке компонентов.
